File: python/periph/transport/smbus.py
from .base import Transport


def _crc8(data):
    crc = 0
    for byte in data:
        crc ^= byte
        for _ in range(8):
            crc = (crc << 1) ^ 0x07 if crc & 0x80 else crc << 1
        crc &= 0xFF
    return crc


class SMBusTransport(Transport):
    """SMBus transport for MicroPython (wraps machine.I2C with address validation and PEC).

    Enforces the valid 7-bit SMBus address range (0x08–0x77) and, when pec=True,
    appends a CRC-8 byte to writes and verifies it on reads.

    Args:
        bus: Configured machine.I2C or machine.SoftI2C instance.
        addr: 7-bit device address (0x08–0x77).
        pec: Enable Packet Error Code (CRC-8) checking (default False).

    Raises:
        ValueError: If addr is outside the valid SMBus range.
    """

    def __init__(self, bus, addr, pec=False):
        if not (0x08 <= addr <= 0x77):
            raise ValueError("SMBus address must be in range 0x08-0x77")
        self._bus = bus
        self._addr = addr
        self._pec = pec

    def write(self, data):
        """Send bytes to the device, appending a PEC byte if enabled.

        Args:
            data: Bytes to write.

        Raises:
            OSError: On no ACK or bus timeout.
        """
        if self._pec:
            data = bytes(data) + bytes([_crc8(bytes([self._addr << 1]) + bytes(data))])
        self._bus.writeto(self._addr, data)

    def read(self, n):
        """Read bytes from the device, verifying the PEC byte if enabled.

        Reads n+1 bytes when PEC is enabled and strips the trailing CRC byte.

        Args:
            n: Number of data bytes to read.

        Returns:
            bytes: The n data bytes (PEC byte stripped if enabled).

        Raises:
            OSError: On PEC mismatch or bus timeout.
        """
        raw = self._bus.readfrom(self._addr, n + 1 if self._pec else n)
        if self._pec:
            if _crc8(bytes([(self._addr << 1) | 1]) + raw[:-1]) != raw[-1]:
                raise OSError("SMBus PEC error")
            return bytes(raw[:-1])
        return bytes(raw)

    def write_read(self, data, n):
        """Write then read with optional PEC verification.

        PEC covers the full transaction: write address + write data +
        read address + read data.

        Args:
            data: Bytes to write (typically a register address).
            n: Number of data bytes to read back.

        Returns:
            bytes: The n data bytes (PEC byte stripped if enabled).

        Raises:
            OSError: On PEC mismatch or bus timeout.
        """
        buf = bytearray(n + 1 if self._pec else n)
        self._bus.writeto_then_readfrom(self._addr, bytes(data), buf)
        if self._pec:
            expected = _crc8(
                bytes([self._addr << 1]) + bytes(data) +
                bytes([(self._addr << 1) | 1]) + bytes(buf[:-1])
            )
            if expected != buf[-1]:
                raise OSError("SMBus PEC error")
            return bytes(buf[:-1])
        return bytes(buf)

File: rust/periph/src/transport/smbus.rs
//! SMBus transport wrapper for chip drivers.
//!
//! Wraps any [`I2c`] implementation and adds 7-bit address validation and
//! optional PEC (Packet Error Code) CRC-8 checking. The wrapper itself
//! implements [`I2c`], so chip drivers generic over [`I2c`] accept an
//! [`SmBusTransport`] directly without knowing about SMBus specifics.
//!
//! ## Usage
//!
//! ```rust,ignore
//! use linux_embedded_hal::I2cdev;
//! use periph::transport::smbus::SmBusTransport;
//!
//! let i2c = I2cdev::new("/dev/i2c-1")?;
//! let smbus = SmBusTransport::new(i2c, 0x40, false)?;
//! // smbus implements I2c — pass it directly to chip driver constructors
//! let chip = MyChip::new(smbus, 0x40);
//! ```
//!
//! ## PEC computation
//!
//! CRC-8 with polynomial `x⁸ + x² + x + 1` (0x07), initial value 0x00.
//!
//! | Operation    | Bytes covered by CRC                                             |
//! |--------------|------------------------------------------------------------------|
//! | `write`      | `(addr << 1)` + data                                            |
//! | `read`       | `(addr << 1) \| 1` + received data                              |
//! | `write_read` | `(addr << 1)` + write data + `(addr << 1) \| 1` + received data |

use core::fmt;
use embedded_hal::i2c::{self, ErrorKind, ErrorType, I2c, Operation};

fn crc8_update(mut crc: u8, byte: u8) -> u8 {
    crc ^= byte;
    for _ in 0..8 {
        crc = if crc & 0x80 != 0 { (crc << 1) ^ 0x07 } else { crc << 1 };
    }
    crc
}

/// Error type returned by [`SmBusTransport`] operations.
#[derive(Debug)]
pub enum SmBusError<E: fmt::Debug> {
    /// Underlying I²C bus error.
    I2c(E),
    /// Received PEC byte did not match the computed CRC-8.
    PecMismatch,
    /// Device address is outside the valid SMBus range (0x08–0x77).
    InvalidAddress,
}

impl<E: i2c::Error + fmt::Debug> i2c::Error for SmBusError<E> {
    fn kind(&self) -> ErrorKind {
        match self {
            Self::I2c(e) => e.kind(),
            Self::PecMismatch | Self::InvalidAddress => ErrorKind::Other,
        }
    }
}

/// SMBus transport wrapper.
///
/// Wraps any [`I2c`] implementation and adds:
/// - 7-bit address validation (0x08–0x77) at construction time
/// - Optional software PEC (CRC-8) on all reads and writes
///
/// Implements [`I2c`] so chip drivers generic over [`I2c`] can accept an
/// `SmBusTransport` directly. The per-call `address` parameter in each
/// [`I2c`] method is used for PEC computation.
pub struct SmBusTransport<I2C> {
    i2c: I2C,
    pec: bool,
}

impl<I2C: I2c> SmBusTransport<I2C> {
    /// Create a new SMBus transport, validating the device address range.
    ///
    /// Returns `Err(SmBusError::InvalidAddress)` if `addr` is outside the
    /// valid SMBus range 0x08–0x77. The address is not stored after
    /// validation; per-call addresses in [`I2c`] methods are used for PEC.
    pub fn new(i2c: I2C, addr: u8, pec: bool) -> Result<Self, SmBusError<I2C::Error>> {
        if addr < 0x08 || addr > 0x77 {
            return Err(SmBusError::InvalidAddress);
        }
        Ok(Self { i2c, pec })
    }

    /// Release the underlying I²C bus, consuming the transport.
    pub fn release(self) -> I2C {
        self.i2c
    }
}

impl<I2C: I2c> ErrorType for SmBusTransport<I2C> {
    type Error = SmBusError<I2C::Error>;
}

impl<I2C: I2c> I2c for SmBusTransport<I2C> {
    fn write(&mut self, address: u8, write: &[u8]) -> Result<(), Self::Error> {
        if self.pec {
            let mut buf = [0u8; 256];
            let len = write.len();
            buf[..len].copy_from_slice(write);
            let mut crc = crc8_update(0, address << 1);
            for &b in write {
                crc = crc8_update(crc, b);
            }
            buf[len] = crc;
            self.i2c.write(address, &buf[..len + 1]).map_err(SmBusError::I2c)
        } else {
            self.i2c.write(address, write).map_err(SmBusError::I2c)
        }
    }

    fn read(&mut self, address: u8, read: &mut [u8]) -> Result<(), Self::Error> {
        if self.pec {
            let n = read.len();
            let mut tmp = [0u8; 256];
            self.i2c.read(address, &mut tmp[..n + 1]).map_err(SmBusError::I2c)?;
            read.copy_from_slice(&tmp[..n]);
            let mut crc = crc8_update(0, (address << 1) | 1);
            for &b in &tmp[..n] {
                crc = crc8_update(crc, b);
            }
            if crc != tmp[n] {
                return Err(SmBusError::PecMismatch);
            }
            Ok(())
        } else {
            self.i2c.read(address, read).map_err(SmBusError::I2c)
        }
    }

    fn write_read(&mut self, address: u8, write: &[u8], read: &mut [u8]) -> Result<(), Self::Error> {
        if self.pec {
            let n = read.len();
            let mut tmp = [0u8; 256];
            self.i2c.write_read(address, write, &mut tmp[..n + 1]).map_err(SmBusError::I2c)?;
            read.copy_from_slice(&tmp[..n]);
            let mut crc = crc8_update(0, address << 1);
            for &b in write {
                crc = crc8_update(crc, b);
            }
            crc = crc8_update(crc, (address << 1) | 1);
            for &b in &tmp[..n] {
                crc = crc8_update(crc, b);
            }
            if crc != tmp[n] {
                return Err(SmBusError::PecMismatch);
            }
            Ok(())
        } else {
            self.i2c.write_read(address, write, read).map_err(SmBusError::I2c)
        }
    }

    fn transaction(&mut self, address: u8, operations: &mut [Operation<'_>]) -> Result<(), Self::Error> {
        self.i2c.transaction(address, operations).map_err(SmBusError::I2c)
    }
}
